sdks/java/core/src/main/java/org/apache/beam/sdk/util/ActiveWindowSet.java

```diff
@@ -181,4 +181,9 @@ void onMerge(Collection<W> toBeMerged, Collection<W> activeToBeMerged, W mergeRe
    * ACTIVE windows {@code toBeMerged} have been merged into {@code mergeResult}.
    */
   W mergedWriteStateAddress(Collection<W> toBeMerged, W mergeResult);
+
+  /**
+   * Return {@literal true} if active window set has any non-empty state.
+   */
+  boolean hasState();
```

Comment:
Sorry, not nuked! This is needed by ReduceFnRunner.emitIfAppropriate. It answers whether there's any active windows which will be persisted at the end of the bundle. It does not need to be ReadableState<Boolean>.